### PR TITLE
Pytest/rtp

### PR DESCRIPTION
**Pull request by mts299**
# Scope 

Adding RTP pytest tests, focuses on just testing the RTP plotting methods and parameters passed in with the test data. However, I still need to add some documentation and imagine the folder structure may need to change but not sure yet. 

I also added some fixes already!

**issue:** #179 

## Approval

**Number of approvals:** 2 - review and tester

## Test

**matplotlib version**: 3.3.4
**Note testers: please indicate what version of matplotlib you are using**

*provide a list or code to test*

```
cd pydarn/test
pytest test_RTP.py 
```

### CODE REVIEW

File: test/test_RTP.py [hunk 1]
```diff
@@ -13,84 +13,177 @@
...
+@pytest.mark.parametrize('cp_name', [True, False])
+@pytest.mark.parametrize('color', 'green')
+@pytest.mark.parametrize('linestyle', ['--'])
+@pytest.mark.parametrize('linewidth', [0.5, 2])
+@pytest.mark.parametrize('start_time', [dt.datetime(2018, 4, 4, 6, 2)])
+@pytest.mark.parametrize('end_time', [dt.datetime(2018, 4, 4, 6, 4)])
+@pytest.mark.parametrize('date_fmt', ['%H:%M'])
+class TestTimeSeries:
+
+    def test_parameters_time_series_channel2(self, parameters_scalar, gate,
```
Comment (carleyjmartin) [inline test/test_RTP.py:113]: First failure on this test: 'color' option on line 105 is taking each of the letters as individual inputs.

File: test/test_RTP.py [hunk 2]
```diff
@@ -13,84 +13,177 @@
...
+                                       cmap=cmap, date_fmt=date_fmt)
+        plt.close('all')
+
+
+@pytest.mark.parametrize('parameters_scalar', ['tfreq', 'cp', 'nave',
+                                               'p_l', 'w_l', 'v'])
+@pytest.mark.parametrize('gate', [38, 48])
+@pytest.mark.parametrize('scale', ['linear', 'log'])
+@pytest.mark.parametrize('cp_name', [True, False])
+@pytest.mark.parametrize('color', 'green')
```
Comment (carleyjmartin) [inline test/test_RTP.py:105]: See comment below 
```suggestion
@pytest.mark.parametrize('color', 'g')
```

File: test/test_RTP.py [hunk 3]
```diff
@@ -13,84 +13,177 @@
...
+    def test_params_summary_plots(self, fig_size, watermark, boundary, cmaps,
+                                  lines, plot_elv, title, background,
+                                  groundscatter, coords):
+        with warnings.catch_warnings(record=True):
+            pydarn.RTP.plot_summary(data, beam_num=15, channel=1,
+                                    figsize=fig_size, watermark=watermark,
+                                    boundary=boundary, cmaps=cmaps,
+                                    lines=lines, plot_elv=plot_elv,
+                                    title=title, background=background,
+                                    groundscatter=groundscatter, coords=coords)
```
Comment (carleyjmartin) [inline test/test_RTP.py:175]: It keeps throwing up that there is an error on this line, but the error doesn't make sense to be coming from this line as its another color issue: 
`ValueError: '.' is not a valid value for color`
The only input that has a '.' is from 'lines'

**Review by carleyjmartin — COMMENTED** [review]
Cloned repo branch, then ran `pytest test_RTP.py`.
Returns: `512 failed, 674 passed, 4 warnings in 255.55s (0:04:15)`
All failures are due to the two comments, something to do with colour values on both.

File: test/test_RTP.py (hunk 1 above)
Comment (mts299) [reply test/test_RTP.py:113]: did you pull this branch down and then recompiled pyDARN?

Comment (mts299) [reply test/test_RTP.py:113]: what version of pytest are you using? Mine does not do that.

Comment (carleyjmartin) [reply test/test_RTP.py:113]: pytest version: pytest-6.2.4

I pulled the branch and then used `python3 setup.py install` to install pydarn, and pulled the branch but installed using `pip3 install git+.......pytest/RTP` too, both with the same results. The error reads (sorry it's long, this is just one but a variation on a theme) this one fails for 'n' but its working for the 'g' and the 'r' so only getting errors relating to 'e's and 'n's:

```
__________________________ TestTimeSeries.test_parameters_time_series_channel2[%H:%M-end_time0-start_time0-2----n-True-log-48-v] __________________________

self = <test_RTP.TestTimeSeries object at 0x128a2a7f0>, parameters_scalar = 'v', gate = 48, scale = 'log', cp_name = True, color = 'n', linestyle = '--'
linewidth = 2, date_fmt = '%H:%M', end_time = datetime.datetime(2018, 4, 4, 6, 4), start_time = datetime.datetime(2018, 4, 4, 6, 2)

    def test_parameters_time_series_channel2(self, parameters_scalar, gate,
                                             scale, cp_name, color, linestyle,
                                             linewidth, date_fmt, end_time,
                                             start_time):
        """
    
        """
        with warnings.catch_warnings(record=True):
            pydarn.RTP.plot_time_series(data, parameter=parameters_scalar,
                                        beam_num=9, gate=gate,
                                        start_time=start_time,
                                        end_time=end_time, date_fmt=date_fmt,
                                        channel=2, scale=scale,
                                        cp_name=cp_name, color=color,
                                        linestyle=linestyle,
>                                       linewidth=linewidth)

test_RTP.py:128: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../venv-pytest/lib/python3.7/site-packages/pydarn-2.1-py3.7.egg/pydarn/plotting/rtp.py:701: in plot_time_series
    linewidth=linewidth)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/__init__.py:1361: in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1676: in plot_date
    return self.plot(x, y, fmt, **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1605: in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:315: in __call__
    yield from self._plot_args(this, kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in _plot_args
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in <listcomp>
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:534: in <genexpr>
    for j, label in enumerate(labels))
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:354: in _makeline
    seg = mlines.Line2D(x, y, **kw)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:374: in __init__
    self.set_color(color)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:1054: in set_color
    mcolors._check_color_like(color=color)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

kwargs = {'color': 'n'}, k = 'color', v = 'n'

    def _check_color_like(**kwargs):
        """
        For each *key, value* pair in *kwargs*, check that *value* is color-like.
        """
        for k, v in kwargs.items():
            if not is_color_like(v):
>               raise ValueError(f"{v!r} is not a valid value for {k}")
E               ValueError: 'n' is not a valid value for color

../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/colors.py:156: ValueError

```


The later error for summary plots is: 
```
____________ TestSummaryPlots.test_params_summary_plots[Coords.RANGE_GATE-False-grey-test test-False-lines0-cmaps0-boundary0-False-fig_size0] _____________

self = <test_RTP.TestSummaryPlots object at 0x129d50128>, fig_size = (11, 10), watermark = False
boundary = {'elv': (0, 20), 'nave': (0, 30), 'tfreq': (0, 15), 'w_l': (0, 100)}
cmaps = {'elv': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>, 'p_l': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>, 'vel': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>}
lines = {'nave': '--', 'tfreq': '.'}, plot_elv = False, title = 'test test', background = 'grey', groundscatter = False, coords = <Coords.RANGE_GATE: 1>

    def test_params_summary_plots(self, fig_size, watermark, boundary, cmaps,
                                  lines, plot_elv, title, background,
                                  groundscatter, coords):
        with warnings.catch_warnings(record=True):
            pydarn.RTP.plot_summary(data, beam_num=15, channel=1,
                                    figsize=fig_size, watermark=watermark,
                                    boundary=boundary, cmaps=cmaps,
                                    lines=lines, plot_elv=plot_elv,
                                    title=title, background=background,
>                                   groundscatter=groundscatter, coords=coords)

test_RTP.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../venv-pytest/lib/python3.7/site-packages/pydarn-2.1-py3.7.egg/pydarn/plotting/rtp.py:956: in plot_summary
    label=labels[i][0], **kwargs)
../../venv-pytest/lib/python3.7/site-packages/pydarn-2.1-py3.7.egg/pydarn/plotting/rtp.py:701: in plot_time_series
    linewidth=linewidth)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/__init__.py:1361: in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1676: in plot_date
    return self.plot(x, y, fmt, **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1605: in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:315: in __call__
    yield from self._plot_args(this, kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in _plot_args
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in <listcomp>
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:534: in <genexpr>
    for j, label in enumerate(labels))
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:354: in _makeline
    seg = mlines.Line2D(x, y, **kw)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:374: in __init__
    self.set_color(color)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:1054: in set_color
    mcolors._check_color_like(color=color)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

kwargs = {'color': '.'}, k = 'color', v = '.'

    def _check_color_like(**kwargs):
        """
        For each *key, value* pair in *kwargs*, check that *value* is color-like.
        """
        for k, v in kwargs.items():
            if not is_color_like(v):
>               raise ValueError(f"{v!r} is not a valid value for {k}")
E               ValueError: '.' is not a valid value for color

../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/colors.py:156: ValueError

```

**Comment by mts299** [comment]
@carleyjmartin I have pytest 5.4.1 so it seems something changed in various versions, we can stick with your suggestion. 

does the `g` change fix all your issues?

**Comment by carleyjmartin** [comment]
> @carleyjmartin I have pytest 5.4.1 so it seems something changed in various versions, we can stick with your suggestion.
> 
> does the `g` change fix all your issues?

I'll just run it again to check with 'g' .... waiting on it running.... yes it solves the first issue!

Still have the second colour issue with line 157, I think, where it's reading in a '.' but wants a colour instead. Error below:


```
____________ TestSummaryPlots.test_params_summary_plots[Coords.RANGE_GATE-False-grey-test test-False-lines0-cmaps0-boundary0-False-fig_size0] _____________

self = <test_RTP.TestSummaryPlots object at 0x129d50128>, fig_size = (11, 10), watermark = False
boundary = {'elv': (0, 20), 'nave': (0, 30), 'tfreq': (0, 15), 'w_l': (0, 100)}
cmaps = {'elv': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>, 'p_l': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>, 'vel': <matplotlib.colors.LinearSegmentedColormap object at 0x10d88a940>}
lines = {'nave': '--', 'tfreq': '.'}, plot_elv = False, title = 'test test', background = 'grey', groundscatter = False, coords = <Coords.RANGE_GATE: 1>

    def test_params_summary_plots(self, fig_size, watermark, boundary, cmaps,
                                  lines, plot_elv, title, background,
                                  groundscatter, coords):
        with warnings.catch_warnings(record=True):
            pydarn.RTP.plot_summary(data, beam_num=15, channel=1,
                                    figsize=fig_size, watermark=watermark,
                                    boundary=boundary, cmaps=cmaps,
                                    lines=lines, plot_elv=plot_elv,
                                    title=title, background=background,
>                                   groundscatter=groundscatter, coords=coords)

test_RTP.py:175: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../venv-pytest/lib/python3.7/site-packages/pydarn-2.1-py3.7.egg/pydarn/plotting/rtp.py:956: in plot_summary
    label=labels[i][0], **kwargs)
../../venv-pytest/lib/python3.7/site-packages/pydarn-2.1-py3.7.egg/pydarn/plotting/rtp.py:701: in plot_time_series
    linewidth=linewidth)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/__init__.py:1361: in inner
    return func(ax, *map(sanitize_sequence, args), **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1676: in plot_date
    return self.plot(x, y, fmt, **kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_axes.py:1605: in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:315: in __call__
    yield from self._plot_args(this, kwargs)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in _plot_args
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:539: in <listcomp>
    return [l[0] for l in result]
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:534: in <genexpr>
    for j, label in enumerate(labels))
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/axes/_base.py:354: in _makeline
    seg = mlines.Line2D(x, y, **kw)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:374: in __init__
    self.set_color(color)
../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/lines.py:1054: in set_color
    mcolors._check_color_like(color=color)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

kwargs = {'color': '.'}, k = 'color', v = '.'

    def _check_color_like(**kwargs):
        """
        For each *key, value* pair in *kwargs*, check that *value* is color-like.
        """
        for k, v in kwargs.items():
            if not is_color_like(v):
>               raise ValueError(f"{v!r} is not a valid value for {k}")
E               ValueError: '.' is not a valid value for color

../../venv-pytest/lib/python3.7/site-packages/matplotlib-3.4.2-py3.7-macosx-10.14-x86_64.egg/matplotlib/colors.py:156: ValueError

```

File: test/test_RTP.py (hunk 3 above)
Comment (carleyjmartin) [reply test/test_RTP.py:175]: Looking further into this, I'm not sure but lines wants colour values, so if I replace '--' and '.' with 'r' and 'g' then it runs with no errors, I guess lines means the line colour and not the line style?

**Review by carleyjmartin — APPROVED** [review]

**Pull request merged**